lib/hourglass.rb

```diff
@@ -1,4 +1,25 @@
-
+# Time complexity: O(n2), to read thru the two dimensions data, it depends on the length of the longest row and the number of rows. So the time compexity would be O(n) * O(m) --> O(n2). (Not sure whether the matrix slice would affect the compexity)
+# Space complexity: O(1), it doesn't depend on the content of the matrix but a fixed number of variables, so the space complexity is O(1).
 def hourglass(matrix)
-  raise NotImplementedError, "The matrix method has not been implemented yet!"
+  start = 0
+  row = 3
+  largest_sum = nil
+  while row < matrix.length + 1
+    col = 3
+    while col < matrix[0].length + 1
+      temp_sum = hourglass_helper(matrix[start...row], col)
```

Comment:
Interesting, just noting this copies 3 of the rows in the matrix.  Could be expensive.

Reply:
That was the part that I was uncertain, but I guess I have to revise the helper method and don't use the slice anymore...orz